src/core/writer.js

```diff
@@ -123,7 +129,55 @@ function computeMD5(filesize, xrefInfo) {
   return bytesToString(calculateMD5(array));
 }
 
-function incrementalUpdate(originalData, xrefInfo, newRefs) {
+function updateXFA(datasetsRef, newRefs, xref) {
+  if (datasetsRef === null || xref === null) {
+    return;
+  }
+  const datasets = xref.fetchIfRef(datasetsRef);
+  const str = bytesToString(datasets.getBytes());
+  const xml = new SimpleXMLParser(/* hasAttributes */ true).parseFromString(
+    str
+  );
+
+  for (const { xfa } of newRefs) {
+    if (!xfa) {
+      continue;
+    }
+    const { path, value } = xfa;
+    if (!path) {
+      continue;
+    }
+    const node = xml.documentElement.searchNode(parseXFAPath(path), 0);
+    if (node) {
+      node.childNodes = [new SimpleDOMNode("#text", value)];
+    } else {
+      warn(`Node not found for path: ${path}`);
+    }
+  }
+  const buffer = [];
+  xml.documentElement.dump(buffer);
+  let updatedXml = buffer.join("");
+
+  const encrypt = xref.encrypt;
+  if (encrypt) {
+    const transform = encrypt.createCipherTransform(
+      datasetsRef.num,
+      datasetsRef.gen
+    );
+    updatedXml = transform.encryptString(updatedXml);
+  }
+  const data =
+    `${datasetsRef.num} ${datasetsRef.gen} obj\n` +
+    `<< /Type /EmbeddedFile /Length ${updatedXml.length}>>\nstream\n` +
+    updatedXml +
+    "\nendstream\nendobj\n";
+
+  newRefs.push({ ref: datasetsRef, data });
+}
+
+function incrementalUpdate(originalData, xrefInfo, newRefs, xref, datasetsRef) {
```

Comment:
Also, the number of parameters here are now large enough to warrant changing this function to take a parameter object instead (at least in my opinion), to aid readability of the call-sites:
```javascript
function incrementalUpdate({
  originalData,
  xrefInfo,
  newRefs,
  xref = null,
  datasetsRef = null,
}) {
```